api/grpcserver/transaction_service_test.go

```diff
@@ -186,3 +195,101 @@ func BenchmarkStreamResults(b *testing.B) {
 		require.Equal(b, maxcount, n)
 	}
 }
+
+type parseExpectation func(tb testing.TB, resp *pb.ParseTransactionResponse, err error)
+
+func expectParseError(code codes.Code, message string) parseExpectation {
+	return func(tb testing.TB, resp *pb.ParseTransactionResponse, err error) {
+		s, ok := status.FromError(err)
+		require.True(tb, ok)
+		assert.Equal(tb, code, s.Code())
+		assert.Contains(tb, s.Message(), message)
+	}
+}
+
+func parseOk() parseExpectation {
+	return func(tb testing.TB, resp *pb.ParseTransactionResponse, err error) {
+		require.NoError(tb, err)
+		require.NotEmpty(tb, resp)
+	}
+}
+
+func TestParseTransactions(t *testing.T) {
+	db := sql.InMemory()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	t.Cleanup(cancel)
+	vminst := vm.New(db)
+	t.Cleanup(launchServer(t, cfg, NewTransactionService(db, nil, nil, txs.NewConservativeState(vminst, db), nil)))
+	var (
+		conn     = dialGrpc(ctx, t, cfg.PublicListener)
+		client   = pb.NewTransactionServiceClient(conn)
+		keys     = make([]signing.PrivateKey, 4)
+		accounts = make([]types.Account, len(keys))
+		rng      = rand.New(rand.NewSource(10101))
+	)
+	for i := range keys {
+		pub, priv, err := ed25519.GenerateKey(rng)
+		require.NoError(t, err)
+		keys[i] = signing.PrivateKey(priv)
+		accounts[i] = types.Account{Address: wallet.Address(pub), Balance: 1e12}
+	}
+	require.NoError(t, vminst.ApplyGenesis(accounts))
+	_, _, err := vminst.Apply(vm.ApplyContext{Layer: types.GetEffectiveGenesis().Add(1)},
+		[]types.Transaction{{RawTx: types.NewRawTx(wallet.SelfSpawn(keys[0], 0))}}, nil)
+	require.NoError(t, err)
+	mangled := wallet.Spend(keys[0], accounts[3].Address, 100, 0)
+	mangled[len(mangled)-1] -= 1
+
+	for _, tc := range []struct {
+		desc   string
+		tx     []byte
+		verify bool
+		expect parseExpectation
+	}{
+		{
+			"empty",
+			nil,
+			false,
+			expectParseError(codes.InvalidArgument, "empty"),
+		},
+		{
+			"malformed",
+			[]byte("something"),
+			false,
+			expectParseError(codes.InvalidArgument, "malformed tx"),
+		},
+		{
+			"not spawned",
+			wallet.Spend(keys[2], accounts[3].Address, 100, 0),
+			false,
+			expectParseError(codes.NotFound, "not spawned"),
+		},
+		{
+			"all good",
+			wallet.Spend(keys[0], accounts[3].Address, 100, 0),
+			false,
+			parseOk(),
+		},
+		{
+			"all good with verification",
+			wallet.Spend(keys[0], accounts[3].Address, 100, 0),
+			true,
+			parseOk(),
+		},
+		{
+			"mangled signature",
+			mangled,
+			true,
+			expectParseError(codes.InvalidArgument, "signature is invalid"),
+		},
+	} {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			resp, err := client.ParseTransaction(context.Background(), &pb.ParseTransactionRequest{
+				Transaction: tc.tx,
+				Verify:      tc.verify,
+			})
+			tc.expect(t, resp, err)
+		})
```

Comment:
i want two things:
- understand test by looking at the inputs and outputs
- minimize amount of code that i have to change if some interface changes. so if i change ParseTransaction - i want to change it in one place, rather than 5

if i type out every test in Run, then i loose both. another alternative that makes sense to me is to make a function called parseTransactionTest and drop table, but it allows more "flexibility" in writing tests, which i would rather avoid when possible